src/python/pants/backend/javascript/goals/test.py

```diff
@@ -0,0 +1,114 @@
+from typing import Any, Iterable
+
+from pants.backend.javascript import install_node_package, nodejs_project_environment
+from pants.backend.javascript.install_node_package import (
+    InstalledNodePackage,
+    InstalledNodePackageRequest,
+)
+from pants.backend.javascript.nodejs_project_environment import NodeJsProjectEnvironmentProcess
+from pants.backend.javascript.subsystems.nodejstest import NodeJSTest
+from pants.backend.javascript.target_types import (
+    JSSourceField,
+    JSTestExtraEnvVarsField,
+    JSTestSourceField,
+    JSTestTimeoutField,
+)
+from pants.core.goals.test import TestExtraEnv, TestFieldSet, TestRequest, TestResult, TestSubsystem
+from pants.core.target_types import AssetSourceField
+from pants.core.util_rules import source_files
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.internals import graph, platform_rules
+from pants.engine.internals.native_engine import Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    SourcesField,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+
+@dataclass(frozen=True)
+class JSTestFieldSet(TestFieldSet):
+    required_fields = (JSTestSourceField,)
+
+    source: JSTestSourceField
+    dependencies: Dependencies
+    timeout: JSTestTimeoutField
+    extra_env_vars: JSTestExtraEnvVarsField
+
+
+class JSTestRequest(TestRequest):
+    tool_subsystem = NodeJSTest
+    field_set_type = JSTestFieldSet
+
+
+@rule(level=LogLevel.DEBUG, desc="Run javascript tests")
+async def run_javascript_tests(
+    batch: JSTestRequest.Batch[JSTestFieldSet, Any],
+    test: TestSubsystem,
+    test_extra_env: TestExtraEnv,
+) -> TestResult:
+    field_set = batch.single_element
+    installation_get = Get(InstalledNodePackage, InstalledNodePackageRequest(field_set.address))
+    transitive_tgts_get = Get(TransitiveTargets, TransitiveTargetsRequest([field_set.address]))
+
+    field_set_source_files_get = Get(SourceFiles, SourceFilesRequest([field_set.source]))
+    target_env_vars_get = Get(
+        EnvironmentVars, EnvironmentVarsRequest(field_set.extra_env_vars.sorted())
+    )
+    installation, transitive_tgts, field_set_source_files, target_env_vars = await MultiGet(
+        installation_get, transitive_tgts_get, field_set_source_files_get, target_env_vars_get
+    )
+
+    sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            (tgt.get(SourcesField) for tgt in transitive_tgts.closure),
+            enable_codegen=True,
+            for_sources_types=[JSSourceField, AssetSourceField],
```

Comment:
The reason for hydrating all `AssetSourceField`s is that I think the easiest way to support the many different configuration files that can play a part in tooling for a node package is to let this 
```BUILD
files(name="configs", sources=["*.rc", "*.json", "*.config.js"])
package_json(
    ...
    dependencies=[":configs"],
)
```
be the way to have them be part of the sandboxes.

Reply:
If a test config is only to affect test caching, I think `__defaults__` configuring `javascript_test`:s dependencies could work?

Reply:
Hmm... I'm not really following the `files` and `cwd` bit fully, but:

I think in pants lingo, any file is considered a `resource` to the `package.json`: All files that will be included in a package .tar are expected to be at or below the directory containing the `package.json`, when "packaging" (npm pack). But the pipeline is scriptable, and the scripts themselves have no limitations to where they can copy files from.

The `package.json` has fields that references where files to include in packages are found, but either silently ignores `../` or errors, depending on package manager. The default is to implicitly include all files next to and below the `package.json`, with a `.gitignore`-esque blocklist.

In source code, files are either loaded via `import "./path/to/file.json"` or via file reading api:s similar to pythons `open` (of course...). Same thing here is that the import cannot "escape" the `package.json` directory, whilst `open` could.

So I guess if a `package_json` happens to have a dependency on a `resource` that isn't at or below it's containing directory, pants could error or warn?

I think `file` is the same, strictly speaking? But that is disregarding scriptability. Users could have scripts in `package.json` to e.g copy files from outside into the `package.json` dir that executes as part of `npm pack`. Pants has facilities for this as well though in `relocated_files`, so maybe encouraging using that is better? :thinking:

Reply:
>That would probably be a good idea until support for source/workspace dependencies is available.

I think that already works, package manager workspaces deals with the module resolution and the dep inference already picks up the source dependency. A workspace package dependendee ends up as a symlink in the `/node_modules` of the the dependent, install time.